Issue 172: MultiList: validate both min and max on keyenter

sveinn opened the pull request:
### Describe the change
Fixes [issue 172](https://github.com/piotrmurach/tty-prompt/issues/172)

### Requirements
<!--- Put an X between brackets on each line if you have done the item: -->
- [ ] Tests written & passing locally?
- [ ] Code style checked?
- [x] Rebased with `master` branch?
- [ ] Documentation updated?
- [ ] Changelog updated?

piotrmurach commented on lib/tty/prompt/multi_list.rb:
```diff
@@ -43,11 +43,12 @@ def max(value)
-        valid = @min <= @selected.size if @min
-        valid = @selected.size <= @max if @max
+        min_valid = true
+        max_valid = true
+        min_valid = @min <= @selected.size if @min
+        max_valid = @selected.size <= @max if @max
 
-        super if valid
+        super if min_valid && max_valid
```
Instead, you could modify the original snippet:

```ruby
valid &= @selected.size <= @max if @max
```

piotrmurach commented on lib/tty/prompt/multi_list.rb:
```diff
@@ -80,11 +82,11 @@ def keyctrl_a(*)
-                    acc
-                  end
-        @selected = SelectedChoices.new(choices.enabled - @selected.to_a, indexes)
+        indexes = choices.each_with_index.each_with_object([]) do |(choice, idx), acc|
+          acc << idx if !choice.disabled? && !@selected.include?(choice)
+        end
+        @selected = SelectedChoices.new(choices.enabled - @selected.to_a,
+                                        indexes)
```
Please refrain from formatting code that isn't affected by the issue.

piotrmurach reviewed (commented):
Thank you for submitting this PR. 

As things stand I'm unable to accept it. At a minimum, there needs to be a test case that proves the issue has been fixed. Also, I'd like you to revert any formatting changes to the code that isn't affected directly by the issue. My preference is for PR to be focused on what is being fixed/added. I'd also appreciate an entry in the changelog.

sveinn commented:
> Thank you for submitting this PR.
> 
> As things stand I'm unable to accept it. At a minimum, there needs to be a test case that proves the issue has been fixed. Also, I'd like you to revert any formatting changes to the code that isn't affected directly by the issue. My preference is for PR to be focused on what is being fixed/added. I'd also appreciate an entry in the changelog.

My pleasure. I've addressed your requests, please let me know if I missed anything. Thanks!

piotrmurach reviewed (commented):
Thank you for working on the changes. It seems you have missed out on a variable that affects some tests. I left few extra comments to keep the tests focused. Whilst reviewing I noticed that it would be good to consider the situation when a user provides min greater than max. Should we check for this and raise an appropriate and meaningful error?

sveinn commented on lib/tty/prompt/multi_list.rb:
```diff
@@ -82,7 +82,6 @@ def keyctrl_r(*)
 
         indexes = choices.each_with_index.reduce([]) do |acc, (choice, idx)|
                     acc << idx if !choice.disabled? && !@selected.include?(choice)
```
Oops, autoformatting bites me again, in an odd way this time.. fixed now.

sveinn commented:
> Thank you for working on the changes. It seems you have missed out on a variable that affects some tests. I left few extra comments to keep the tests focused. Whilst reviewing I noticed that it would be good to consider the situation when a user provides min greater than max. Should we check for this and raise an appropriate and meaningful error?

Woops, sorry about that. 

> Should we check for this and raise an appropriate and meaningful error?

Good idea. Made a stab at this, let me know what you think.

piotrmurach commented on spec/unit/multi_select_spec.rb:
```diff
@@ -905,4 +909,48 @@ def exit_message(prompt, choices)
+        expect(prompt.output.string).to eq(expected_output)
+      end
+    end
+    context "with min > max" do
```
Please remove nested context, no need for this, it is enough to have a clear test description.

piotrmurach commented on spec/unit/multi_select_spec.rb:
```diff
@@ -905,4 +909,48 @@ def exit_message(prompt, choices)
+      end
+    end
+    context "with min > max" do
+      let(:min) { 3 }
+      let(:max) { 2 }
```
Unnecessary lets, in general, I'm not a fan of `lets`, use them sporadically. Use the integers directly in the test body without any variable assignment.

piotrmurach commented on spec/unit/multi_select_spec.rb:
```diff
@@ -905,4 +909,48 @@ def exit_message(prompt, choices)
+    context "with min > max" do
+      let(:min) { 3 }
+      let(:max) { 2 }
+      it "raises an error message" do
```
I much prefer a direct explanation in the test description: "raises an error message when the min is greater than max".

piotrmurach commented on spec/unit/multi_select_spec.rb:
```diff
@@ -905,4 +909,48 @@ def exit_message(prompt, choices)
+      it "raises an error message" do
+        options = {min: min, max: max, per_page: choices.count}
+        error_msg = "min must not be greater than max"
+        expect { prompt.multi_select("What letter?", choices, options) }.to raise_error(TTY::Prompt::ConfigurationError, error_msg)
```
I'd split the assertion on multiple lines and inline variables and remove unnecessary `per_page` as it defaults to 6 items anyhow. So the test would be reduced to the following:

```ruby
choices = %w[A B C]
expect { 
  prompt.multi_select("What letter?", choices, min: 3, max: 2) 
}.to raise_error(TTY::Prompt::ConfigurationError,
                 "min must not be greater than max")
```

piotrmurach commented on spec/unit/multi_select_spec.rb:
```diff
@@ -905,4 +909,48 @@ def exit_message(prompt, choices)
   end
+
+  context "with :min and :max" do
+    let(:choices) { %w[A B C] }
+    before do
+      prompt.on(:keypress) do |e|
+        prompt.trigger(:keydown) if e.value == "j"
+      end
+    end
+    context "with min <= max" do
+      let(:min) { 2 }
+      let(:max) { 2 }
```
Please remove nested context and inline all the `let` inside the test body as well as the `before`. I'd also use the min and max directly as integers in the `prompt.multi_select` call.

piotrmurach commented on spec/unit/multi_select_spec.rb:
```diff
@@ -905,4 +909,48 @@ def exit_message(prompt, choices)
+      let(:min) { 2 }
+      let(:max) { 2 }
+      it "requires a min number of choices" do
+        keydown = "j"
+        spacebar = " "
+        enter = "\r"
```
Let's follow the other tests and inline these as inputs. I don't think it will distract from test meaning if we do:

```ruby
prompt.input << " " << "\n" << "j" << " " << "\n"
```

piotrmurach commented on lib/tty/prompt/multi_list.rb:
```diff
@@ -23,6 +23,8 @@ def initialize(prompt, **options)
         @echo = options.fetch(:echo, true)
         @min  = options[:min]
         @max  = options[:max]
+        error = "min must not be greater than max"
+        raise ConfigurationError, error if @min && @max && @min > @max
```
I'd extract to a separate private method like `check_minmax` or similar like so:

```ruby
# @api private
def check_minmax
  return unless @min && @max && @min > @max

  raise ConfigurationError, "min must not be greater than max"
end
```

Then you can reuse this method for the `min` and `max` methods right after the initializer.

sveinn replied:
I took this to mean that instead of invoking this method from the initializer that we should invoke it from both the `min` and the `max` setter methods. I wasn't sure exacly where in the file you'd like this method defined; I opted for in the `private` section, close to the `minmax_help` method.

piotrmurach reviewed (commented):
Thanks for addressing the suggestions so quickly. We're nearly there. I left few more comments on the min and max checking as well as tests formatting.

sveinn commented:
> Thanks for addressing the suggestions so quickly. We're nearly there. I left few more comments on the min and max checking as well as tests formatting.

Sure thing. Just addressed your comments. Lemme know if there are any more. Thanks!

piotrmurach reviewed (commented):
Things are looking good, I left few minor comments.

sveinn commented on spec/unit/multi_select_spec.rb:
```diff
@@ -17,8 +17,12 @@ def output_helper(prompt, choices, active, selected, options = {})
     out << prompt << " "
-    out << "(min. #{options[:min]}) " if options[:min]
-    out << "(max. #{options[:max]}) " if options[:max]
+    if options[:min] && options[:max]
+      out << "(min. #{options[:min]}, max. #{options[:max]}) "
+    else
+      out << "(min. #{options[:min]}) " if options[:min]
+      out << "(max. #{options[:max]}) " if options[:max]
+    end
```
If we're optimizing for readability/ease of change this feels less optimal to me since the current version makes it clear that there are two distinct cases and the outcome is conspicuous for each case -- don't have to slow down to compute that mentally. But if happy to change if you feel strongly.

piotrmurach reviewed (commented):
Thank you for such a quick turnaround. I left a comment on the changelog description. One thing to make you aware of is that I'm planning to merge another PR before yours as it was long in waiting and it is a significant change. There shouldn't be any conflicts with your work but you'll need to rebase.

sveinn commented on CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
+
+### Fixed
+* Fix MultiList#keyenter validation - min option also validated now when both min and max are specified. 
+* Fix MultiList#keyenter validation - ensure min is not greater than max
```
Made a change and re-requested a review.

Thomascountz commented:
👋🏾 Hi there, is there any additional help needed on this PR? I was just about to open [a new one](https://github.com/Thomascountz/tty-prompt/pull/1), but I'd love to help get this one merged instead 😃
